### ISSUE TITLE
Patching submodules doesn't work

### ISSUE THREAD

**Issue by robintw** [recipy/recipy#52]
Bug report from @MichielCottaar (via PR):

>The problem is that the submodules are not imported automatically, so for example there is a gifti submodule containing GiftiImage. However if I add "gifti.GiftiImage" to the list of images in my patch, I get an AttributeError:
AttributeError: 'module' object has no attribute 'gifti'
So the patching of GiftiImage would have to happen not when import nibabel is run, but when from nibabel import gifti is run. So to support these additional data formats I see two options:
1. convince the maintainers of nibabel to import all submodules in the main init.py
2. write an additional PatchGifti to patch GiftiImage, when from nibabel import gifti is run. However, I would be a bit uncomfartable with the second option without some way to check that the gifti being loaded is in fact from the nibabel module. I have no idea if that is possible...

We need to try and solve this somehow!

**Comment by MichielCottaar** [comment]
When I uncommented the print-function in `PatchImporter.find_module` I found that if you run `from nibabel import gifti` the `module_name` is still set to the complete name ("nibabel.gifti"). So a possible solution for my problem would be to define a seperate `PatchSimple` for every submodule which defines i/o. The i/o functions would then get patched no matter if it is imported as `import <module>` or `from <module> import <sub_module>`.

Unfortunately this would involve writing a lot of PatchSimple subclasses. I guess this should be possible through metaclasses, but I was wondering if there is a reason why the `module_name`, `input_functions`, `input_wrappers` etc. can not be just set through the __init__ method rather than in the class definitions? That would make generating PatchSimple instances for every submodule very simple...

**Comment by robintw** [comment]
Yes, doing it through `__init__` methods would also be possible. The reason I did it as a full subclass originally was so that methods could be overriden if necessary (I wasn't sure how complicated it would become to deal with importing).

I can see two ways forward:

1) We do what you suggested and define a new class (`PatchStandard` maybe?) which can have these parameters set in the call to `__init__`.

2) We see if we can define a new more complicated subclass of `PatchImporter` which deals with this internally, by defining multiple module names (including the subpackages) and then patching the required ones whenever they appear... This might be cleaner in the long-term, but will be a bit more difficult.

**Comment by MichielCottaar** [comment]
skimage also suffers from this bug, because the io submodule is not imported automatically.... 
Your second solution would be complicated, because we would have to automatically distinguish between methods defined in submodules (e.g. io.imread) and those defined in the main module (e.g. Image.open). I can think of two solutions to bring this to work:
1. We could alter the `find_module` method to attempt to patch the function when either the main module or any submodule is imported and silently catch any AttributeErrors raised. This would require keeping track of which functions/methods have already been successfully patched to prevent them from getting patched again in subsequent imports. The drawback of this solution would be that any spelling errors in the names of the functions to be patched would be silently passed over and not raise any errors.
2. Alternatively we could simply import all the functions listed in input_functions and output_functions before patching them. This would ensure that these functions are actually available, but it would be an override of the lazy importing behaviour of some modules (which might be a problem for e.g. matplotlib, which does a lot during the import of the pyplot submodule).
If you can live with these drawbacks, I'll be happy to implement either of these.

**Comment by robintw** [comment]
Hmm, I'm not sure what's best really. It might be best to go back to your original idea of `__init__` methods, and see if we can add some syntactic sugar to make it as easy as possible for people to use it.

I'm off to another conference next week (yes, I know!) but will have a think and see what I can do. If you want to start working on a piece of code I'd be more than happy to look at it and discuss it with you - I'm just not sure what the best path to go down is, and don't want you to waste your time.

**Comment by robintw** [comment]
Just a few notes for myself here:

I've tried implementing `PatchVerySimple`, but there seems to be some issue with instance variables and class variables: but I can't understand how my code might _require_ class variables....weird!
